### DISCUSSION TITLE
[Feature request] User configurable sensor configuration

### DISCUSSION THREAD

**Discussion by platisd** [ItJustWorksTM/smce-gd#41] [Ideas]
How big of an issue is having the user being able to configure sensors on the car?

No graphical representation of how the sensors would look like is necessary. Potentially even providing a JSON file would suffice.
The only "restriction" would be that sensors can be placed in set positions (front-right, front, front-left, middle-right, middle-left, rear-right, rear, rear-left) and that the pins they correspond to shouldn't be tied to something else already (and be valid).

**Comment by RuthgerD** [comment]
So the most feasible thing is this:
Only allow the config to modify existing attachments (as predefined in the car scene).
As a con this would still disallow you to specify if you want a Ultrasonic or an Infrared.
As a pro this allows the car implementation to directly use the attachments (and the other way round), for example the Motor attachments produce a throttle value which the car then uses to do all the physics stuff to make the car move; this is kind of impossible to do if both sides aren't allowed to make assumptions about existence.

So we could make a hybrid system, where say you have free "slots" where you can put any attachment that doesn't depend on anything but then also still incorporate the modify existing attachments model to still configure built ins.

here we say we want a raycast on the front, and then configure it along with modifying the built in left odometer:
```json
{
   "slots":{
      "front":{
         "class":"AnalogRaycastGD",
         "name":"Front Infrared"
      },
   },
   "config":{
      "Front Infrared":{
         "pin":33,
         "min_distance":4
      },
      "Left Odometer":{
         "distance_pin":35,
         "direction_pin":34
      }
   }
}
```

Ill end my ramblings here for now (Im just making it up on the spot), let me know what you think of this approach.
if this is something we want we can discuss the UX implications of this.

**Reply by AeroStun** [reply]
+1 for "free predefined slots" on which any attachment with no dependency can be bound. Seems like the most versatile approach.

**Reply by platisd** [reply]
As discussed, what I meant was the _hybrid_ approach. I think we will eventually be asked to have different permutations of the **sensor** configuration, so it will be easier to let that up to the user.
In other words, they will be able to define the sensors and the slot, but not things such as motors or odometers.

**Reply by RuthgerD** [reply]
Alright then everyone knows what to expect, the only thing I have yet to decide on is how the user supplies their config file;

the easiest way is to just have a single json file in the user directory which people can modify, this is not very modular but easy

an easy "modular" way but crappy ux is have the user select the config file on every sketch open (will be super annoying)

ideally we want to have some persistent data with an internal configuration file such that the user only selects their config only one time in the ui itself, but that requires implementing a bunch of generic infrastructure to properly support user settings (cuz if we implement persistent data might as well go all the way) might work on that over the "vacation" we have

**Reply by platisd** [reply]
I don't know if that's what you meant, but what about a json file in the sketch folder?
If they don't supply one then a default config is provided. If they have multiple sketches running they will be able to have different cars with different configurations.
